lib/internal/modules/esm/loader.js

```diff
@@ -804,13 +815,22 @@ class ESMLoader {
       hookName: 'resolve',
       shortCircuited: false,
     };
-
     const context = {
       conditions: DEFAULT_CONDITIONS,
       importAssertions,
       parentURL,
     };
-    const validate = (hookErrIdentifier, { 0: suppliedSpecifier, 1: ctx }) => {
+
+    const validate = (hookErrIdentifier, output) => {
+      if (output instanceof Promise) { // eslint-disable-line node-core/prefer-primordials
+        throw ERR_INVALID_RETURN_VALUE(
+          'an object',
+          hookErrIdentifier,
+          output,
+        );
+      }
+
+      const { 0: suppliedSpecifier, 1: ctx } = output;
```

Comment:
I don't think we should do that, `Promise` instances are also objects, and the other checks are enough to catch this case anyway.
```suggestion
    const validate = (hookErrIdentifier, { 0: suppliedSpecifier, 1: ctx }) => {
```